File: dotnet/examples/Pactman.NonprofitCheckPlus.Examples/Cases/Ex25RawAndForwardCompat.cs
using System;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;
using Pactman.NonprofitCheckPlus.Dev;

namespace Pactman.NonprofitCheckPlus.Examples.Cases;

/// <summary>
/// EX-25 — Raw response and forward compatibility.
/// </summary>
/// <remarks>
/// A response from a newer API version than this SDK: unknown fields, an unknown member
/// inside a known object, and an enum value outside the documented set. None of it is an
/// error, and none of it is dropped.
/// </remarks>
public sealed class Ex25RawAndForwardCompat : IExample
{
    /// <inheritdoc />
    public string Id => "ex-25";

    /// <inheritdoc />
    public string Title => "Fields newer than this SDK, readable without an upgrade.";

    /// <inheritdoc />
    public async Task RunAsync(string[] args)
    {
        using var context = ExampleContext.WithFixtures();

        var result = await context.Client.Nonprofits.CheckAsync(Fixtures.Eins.FutureFields);
        var organization = result.Nonprofit!;

        Output.Heading("The record parsed without complaint");
        Output.Field("status", result.Status);
        Output.Field("organization_name", organization.OrganizationName);
        Output.Field("fields returned", organization.Count);

        // What this SDK version predicts, read from the contract shipped inside it.
        var known = Fixtures.KnownNonprofitFields().ToHashSet(StringComparer.Ordinal);
        var unknown = organization.FieldNames.Where(field => !known.Contains(field)).ToList();

        Output.Heading("Fields newer than this SDK");

        foreach (var field in unknown)
        {
            Output.Field(field, organization.Get(field));
        }

        if (unknown.Count == 0)
        {
            Output.Bullet("None — this deployment matches what the SDK predicts.");
        }

        Output.Heading("Reading them");
        Output.Field("Has(...)", organization.Has("state_charity_registration_status"));
        Output.Field("GetString(...)", organization.GetString("state_charity_registration_status"));

        // A nested object a future version added is readable in whatever shape suits.
        var screening = organization.GetElement("watchlist_screening");

        if (screening is { ValueKind: JsonValueKind.Object })
        {
            Output.Field("watchlist_screening.provider", screening.Value.GetProperty("provider").GetString());
            Output.Field("watchlist_screening.matches", screening.Value.GetProperty("matches").GetRawText());
        }

        Output.Heading("An unknown member inside a known object");

        foreach (var entry in organization.OrganizationTypes)
        {
            Output.Field("deductibility_status_description", entry.DeductibilityStatusDescription);
            Output.Field("future_deductibility_note", entry.GetString("future_deductibility_note"));
        }

        Output.Heading("An enum value outside the documented set");
        Output.Field("foundation_type_code", organization.FoundationTypeCode);
        Output.Field("this application would", organization.FoundationTypeCode switch
        {
            "pc" => "route as a public charity",
            "pf" => "route as a private foundation",

            // Never a friendly default. An unrecognized classification is a reason to
            // ask a human, not to guess.
            _ => "route to review — the code is not one this application maps",
        });

        Output.Heading("The whole envelope, unmodified");
        Output.Field("Raw.IsJson", result.Raw.IsJson);
        Output.Field("Raw keys", string.Join(", ", result.Raw.Envelope.EnumerateObject().Select(p => p.Name)));

        Output.Note(
            "The model keeps the wire fields as its store and layers typed properties over\n"
            + "them, so a field the API adds tomorrow is readable today. Log the unknown\n"
            + "ones — they are how you find out the API moved.");
    }
}

File: ruby/examples/ex_25_raw_and_forward_compat.rb
# frozen_string_literal: true

# EX-25 — Raw response and forward compatibility.
#
# The model and the raw envelope are both available, and the raw one is not a
# debugging afterthought. When the API adds a field, the gem you have installed
# keeps working and the new field is readable immediately — no upgrade, no
# deserialization failure, no dropped data.
#
# The fixture used here is an approved response from a newer API version. It
# carries fields this SDK has never heard of and an enum value outside the
# documented set.
#
# Run:  PACTMAN_API_KEY=... bundle exec ruby examples/ex_25_raw_and_forward_compat.rb

require_relative "lib/fixture_api"
require_relative "lib/print"

NCP = Pactman::NonprofitCheckPlus
include Print

# Documented foundation types. Anything else is unknown, not wrong.
KNOWN_FOUNDATION_TYPES = %w[pc pf po].freeze

FixtureApi.with_fixture_api do |client|
  result = client.nonprofits.check(FixtureApi::EINS[:future_fields])
  nonprofit = result.nonprofit

  if nonprofit.nil?
    puts "No record returned."
    next
  end

  bmf = NCP::Sources.bmf(nonprofit)
  pub78 = NCP::Sources.pub78(nonprofit)

  # Known fields read exactly as they always have.
  heading "Known fields are unaffected"
  field "ein", nonprofit.ein
  field "organization_name", nonprofit.organization_name
  field "bmf_status", returned(bmf, :status)
  field "pub78_verified", returned(pub78, :verified)
  field "subsection_description", returned(bmf, :subsection_description)

  # Unknown fields ride along on the same object. No cast, no upgrade.
  unknown_fields = nonprofit.keys - FixtureApi.known_nonprofit_fields

  heading "Fields this SDK version does not declare"
  field "count", unknown_fields.size

  unknown_fields.each { |key| bullet "#{key} = #{JSON.generate(nonprofit[key])}" }

  # Declared fields have readers; undeclared ones are reached with `[]`, which
  # returns whatever the JSON held, so check what you got before you use it.
  registration = nonprofit["state_charity_registration_status"]

  field "read with []", registration.is_a?(String) ? registration : "<not a String>"
  field "responds to a reader", nonprofit.respond_to?(:state_charity_registration_status)

  # An unknown value in a known field. This is the one that breaks applications
  # that map eagerly into an enum and default the miss.
  heading "An unrecognized value in a documented field"

  foundation_type = bmf&.foundation_type_code

  field "foundation_type_code", foundation_type
  field "in the documented set", KNOWN_FOUNDATION_TYPES.include?(foundation_type)
  field "foundation_type_description", returned(bmf, :foundation_type_description)
  field "handled as",
        if KNOWN_FOUNDATION_TYPES.include?(foundation_type)
          "a known classification"
        else
          "unknown — routed to review, not defaulted to a known type"
        end

  # Nested objects keep their unknown members too.
  first_type = pub78&.organization_types&.first

  heading "Unknown members inside a known object"
  field "deductibility_limitation", returned(first_type, :deductibility_limitation)
  field "deductibility_status_description", returned(first_type, :deductibility_status_description)
  field "future_deductibility_note", returned(first_type, :future_deductibility_note)

  # And the whole envelope, as parsed.
  heading "The raw envelope"
  field 'raw["code"]', result.raw["code"]
  field 'raw["message"]', result.raw["message"]
  field 'raw["timeTaken"]', result.raw["timeTaken"]
  field 'raw["nonprofit_check_count"]', result.raw["nonprofit_check_count"]
  field 'raw["data"] is the record', result.raw["data"].equal?(nonprofit.to_h)
  field "top-level envelope keys", result.raw.keys.join(", ")

  bullet "Persist `raw` when you need to prove later what the API actually said."
  bullet "It is the parsed body, unmodified — nothing was dropped on the way through."
end

note "Forward compatibility cuts both ways: an unknown value must never be coerced\n" \
     "into a known one. \"I do not recognize this\" is a valid, and usually safer,\n" \
     "outcome than a confident wrong answer."
